File: allcode/3500-3599/3507minimumPairRemoval.py
# 给你一个数组 nums，你可以执行以下操作任意次数：
#
# 选择 相邻 元素对中 和最小 的一对。如果存在多个这样的对，选择最左边的一个。
# 用它们的和替换这对元素。
# 返回将数组变为 非递减 所需的 最小操作次数 。
#
# 如果一个数组中每个元素都大于或等于它前一个元素（如果存在的话），则称该数组为非递减。
#
#
#
# 示例 1：
#
# 输入： nums = [5,2,3,1]
#
# 输出： 2
#
# 解释：
#
# 元素对 (3,1) 的和最小，为 4。替换后 nums = [5,2,4]。
# 元素对 (2,4) 的和为 6。替换后 nums = [5,6]。
# 数组 nums 在两次操作后变为非递减。
#
# 示例 2：
#
# 输入： nums = [1,2,2]
#
# 输出： 0
#
# 解释：
#
# 数组 nums 已经是非递减的。
#
#
#
# 提示：
#
# 1 <= nums.length <= 50
# -1000 <= nums[i] <= 1000

from leetcode.allcode.competition.mypackage import *

class Solution:
    def minimumPairRemoval(self, nums: List[int]) -> int:
        def check(arr):
            # print(list(p <= q for p, q in pairwise(arr)))
            return all(p <= q for p, q in pairwise(arr))
        ans = 0
        while not check(nums):
            print(nums)
            pair = list(q + p for p, q in pairwise(nums))
            mn = min(pair)
            for i, x in enumerate(pair):
                if x == mn:
                    break
            nums = nums[:i] + [pair[i]] + nums[i + 2:]
            ans += 1
        return ans


so = Solution()
print(so.minimumPairRemoval([-2,1,2,-1,-1,-2,-2,-1,-1,1,1]))  # 10
print(so.minimumPairRemoval([5,2,3,1]))
print(so.minimumPairRemoval([1,2,2]))

File: c++/src/3000-3999/3507minimumPairRemoval.cpp
#include "lc_pub.h"


class Solution {
    public:
    int minimumPairRemoval(vector<int>& nums) {
        auto check=[](vector<int> &arr) -> bool {
            int m=arr.size();
            for (int i=0;i<m-1;i++) {
                if (arr[i]>arr[i+1]) return false;
            }
            return true;
        };

        int ans=0;
        while (!check(nums)) {
            int m=nums.size();
            vector<int> arr(m-1,0);
            int mn=1e5, pos=0;
            for (int i=0;i<nums.size()-1;i++) {
                int s = nums[i]+nums[i+1];
                if (s < mn) {
                    mn=s;
                    pos=i;
                }
            }
            for (int i=0;i<pos;i++) arr[i]=nums[i];
            arr[pos]=nums[pos]+nums[pos+1];
            for (int i=pos+2;i<nums.size();i++) arr[i-1]=nums[i];
            swap(nums,arr);
            ans++;
        }
        return ans;
    }
    };

    
int main()
{
    cout<<"test let us start! %s" << __cplusplus <<std::endl;
    auto grid=parseGrid("[[1,21]]");
    vector<int> nums{5,2,3,1};

    Solution so;
    cout<<so.minimumPairRemoval(nums)<<endl;
    return 0;
}
